### ISSUE TITLE
Question regrading normalization in sinkhorn-knopp

### ISSUE THREAD

**Issue by kkp15** [facebookresearch/capi#2]
Hi, thank you for sharing the code!

I was wondering why the reduce_sum here was done only on gpu-axis (torch.distributed.all_reduce(summed)) but not in batch-axis.
https://github.com/facebookresearch/capi/blob/980a335b6cad0b3da6f97c0004a20d50aaee52c5/model.py#L521

Given that images are sampled independently, I assume it shouldn't hurt doing the reduce_sum also in batch-axis?
Not sure I'm thinking correctly.
Thank you!

**Comment by TimDarcet** [comment]
Hi, thanks for your question!
The SK function is meant to take any tensor `M` of shape "*b n p", and do the SK on each `n*p` matrix independently, while leaving dimensions `*b` unaffected.
When called by the online clustering during the training, **the matrix M has shape `num_patches` x `batch_size` x `num_prototypes`**. This is dues to the transpose [here](https://github.com/facebookresearch/capi/blob/980a335b6cad0b3da6f97c0004a20d50aaee52c5/train_capi.py#L99).
So what happens is that the SK is done independently at each location, ie it's the "positionwise SK" or "modified SK" proposed in the paper.
Tell me if that clarifies, I may have misunderstood the question.

**Comment by kkp15** [comment]
Hello,

Thank you very much for your prompt response. I realized that I had misunderstood the code—I initially assumed that *b n p referred to (batch, spatial/sequence, num_prototypes). Now that I understand it correctly, everything makes sense.

I do have one more question: In this scenario, do you think the batch size plays a significant role, especially since we are using positionwise SK? Additionally, how would you recommend handling data with varying sequence lengths, considering that the longest sequences might effectively reduce the batch size (for example, some patches in a batch with long sequence length could result in an effective batch size of one)?

Thank you again for your help!

**Comment by kkp15** [comment]
Thank you for the kind explanation :)

**Comment by kkp15** [comment]
Sorry for nagging you again.

There's something I don't understand very well regarding the stable_exp part.

The resulting shape of the `shift` will be (b, 1, p) which is `num_patches x 1 x num_prototype`

In extreme cases, and in extreme cases let's say the `num_patches x 1 x num_prototype` came from the first element of the batch. 

In this case, the first element of this batch will going to be all zero, which I think is not what we want.

Am I missing something?....

def stable_exp(M: Float[Tensor, "*b n p"]) -> Float[Tensor, "*b n p"]:
    shift = M.max(dim=-2, keepdim=True).values
    if torch.distributed.is_initialized():
        torch.distributed.all_reduce(shift, torch.distributed.ReduceOp.MAX)
    M += exp_max_values[M.dtype] - shift
    return M.exp()

**Comment by TimDarcet** [comment]
No problem.

The idea of this stable_exp is that right after it, we will divide the matrix's rows by their sum.
That means that multiplying each element of a row by a constant will not affect the result of the algorithm.
Here we add a constant to each element of a row right before exponentiating them, so we are effectively myltiplying this row by a constant, and it will not affect results.
This exp-->divide process is also a softmax, and as you may know softmax is invariant to additive shifts


The reason we do all this is to ensure we do not overflow: if an element has value 100, its value after exp will not fit in a float

**Comment by TimDarcet** [comment]
Thanks for the example, I think I understand better. You are correct, the relative magnitude of logits is not preserved across prototypes. This would happen even without the stable exp part, right after the first division in the SK. It is due to the choice of starting the SK with a first normalization along the batch, which is inherited from DINOv2 and SwAV (https://github.com/facebookresearch/swav/blob/06b1b7cbaf6ba2a792300d79c7299db98b93b7f9/main_swav.py#L354). Starting with a normalization along prototypes works too IIRC.

Intuitively, the explanation is that if all logits of assignment to prototype 1 are significantly greater than to prototype 2, we do wish to reduce the logits of prototype 1 to balance out the assignments. That is why the SK does that, and why it is there.

**Comment by kkp15** [comment]
Yeah, so I was also thinking of doing the normalization along the prototype axis first too but this would result in not doing normalization on the prototype axis in the final normalization step, and the sum of the prototype axis is not guranteed to be 1.

We could just do one more normalization step along the prototype axis after all iterations as a final step. But I'm not very confident if this is a sound approach to do in theory.

**Comment by kkp15** [comment]
This seems more stable. You could do row normalization first but requires more iteration to make it 1, but I think it's more a natural way of doing normalization although it requires more computation.

Let me know what you think!

```
def log_sinkhorn_knopp(M, n_iterations=3):

    # Start directly with log-space values (input matrix)
    log_P = M.clone()
    
    for i in range(n_iterations):

        # Column normalization in log-space
        log_col_sum = torch.logsumexp(log_P, dim=0, keepdim=True)
        log_P = log_P - log_col_sum
        
        # Row normalization in log-space
        log_row_sum = torch.logsumexp(log_P, dim=1, keepdim=True)
        log_P = log_P - log_row_sum

    
    # Convert back to probability space at the end
    return torch.exp(log_P)
```

**Comment by TimDarcet** [comment]
Doing the SK in log-space is a nice idea! It should probably work as far as I can tell.

About the order of normalizations: I'm not very familiar with the convergence proofs, but you should know the algorithm never really converges anyway, 3 iterations is too little for that. This order of normalizations sounds reasonable to me, but I guess only experiments can tell you whether it is sound.

**Comment by kkp15** [comment]
It seems stable now :)
